src/core/observer/index.js

```diff
@@ -219,11 +219,11 @@ export function set (target: Array<any> | Object, key: any, val: any): any {
  * Delete a property and trigger change if necessary.
  */
 export function del (target: Array<any> | Object, key: any) {
-  if (Array.isArray(target)) {
+  if (Array.isArray(target) && typeof key === 'number') {
     target.splice(key, 1)
     return
   }
-  const ob = target.__ob__
+  const ob = (target : any).__ob__
```

Comment:
I can't say I understand it either, but without this cast Flow complains that "Array doesn't have an
__ob__ property" -- why it decided that `target` is an array, I don't know.  It appears to be an instance of [this Flow issue](https://github.com/facebook/flow/issues/1330), though my knowledge of Flow is minimal so I can't tell for sure.